scripts: tests: Handler

LukaszMrugala opened the pull request:
Expansion of the tests related to handlers.py.

Those tests were aimed to improve coverage of the twister tests, thus they want to explore as many of the code branches as possible. One function has been created per method tested, in order to reduce the setup and teardown code necessary. Extensive mocking is used to isolate the methods as much as it is possible.

LukaszMrugala committed:
scripts: tests: Handler

Expansion of the tests related to handlers.py
Achieves over 85% coverage.
Implemented most of golowanow's suggestions.

Signed-off-by: Lukasz Mrugala <lukaszx.mrugala@intel.com>

golowanow commented on scripts/tests/twister/test_handlers.py:
```diff
@@ -0,0 +1,815 @@
+        assert not any([r.levelname == 'ERROR' for r in caplog.records])
+    else:
+        assert len(caplog.text.lower()) > 0
+        assert 'timeout' not in caplog.text.lower()
```
why it checks for absence of `timeout` instead of `timed out`  ? 
also it should check for `Custom script failure` logged here.

golowanow commented on scripts/tests/twister/test_handlers.py:
```diff
@@ -0,0 +1,815 @@
+
+    mock_process.communicate = raise_timeout_fn
+
+    script = mock.Mock()
```
i think it might be just a string here to check it later in the `caplog` when it is expected.

golowanow commented on scripts/tests/twister/test_handlers.py:
```diff
@@ -0,0 +1,815 @@
+        DeviceHandler.run_custom_script(script, timeout)
+
+    if raise_timeout:
+        assert 'timeout' in caplog.text.lower() or 'timed out' in caplog.text.lower()
```
could you please check that `communicate()` was called twice in this case, after `kill()` ?

golowanow reviewed (commented):
@LukaszMrugala, great work!
so far I have time to look only through some of `DeviceHandler` tests.

golowanow commented on scripts/tests/twister/test_handlers.py:
```diff
@@ -0,0 +1,822 @@
+TESTDATA_7 = [
+    (0, False),
+    (4, False),
+#  (0, True) # Known bug
```
is it already fixed with #60206  ?

golowanow commented on scripts/tests/twister/test_handlers.py:
```diff
@@ -0,0 +1,822 @@
+    instance.platform = platform
+    type(instance.testsuite).timeout = mock.PropertyMock(return_value=60)
+    type(instance.platform).timeout_multiplier = mock.PropertyMock(return_value=2)
+
```
```suggestion
type(instance).status = None
type(instance).reason = "Unknown"
```

LukaszMrugala replied:
`type(instance).member = ...` syntax is necessary only when we want to have a `mock.PropertyMock`.
What is the reason for this change?

golowanow replied:
> `type(instance).member = ...` syntax is necessary only when we want to have a `mock.PropertyMock`. What is the reason for this change?

as a suggestion, m.b. we'll need a mock property for some other test ?

LukaszMrugala replied:
Sadly, once you `mock.PropertyMock` something, it is not reassignable. See example below:
```python
a = mock.Mock()
type(a).member = mock.PropertyMock(return_value='Mocked')
print(a.member)  # Prints 'Mocked'
a.member = 'Not mocked'
print(a.member)  # ! Still prints 'Mocked'
```

Because of that, `mock.PropertyMock`ing `status` and/or `reason` breaks the tests.

golowanow commented on scripts/tests/twister/test_handlers.py:
```diff
@@ -0,0 +1,822 @@
+        hardware = handler.get_hardware()
+
+    if raise_exception:
+        assert 'dummy message' in caplog.text.lower()
```
```suggestion
        assert 'dummy message' in caplog.text.lower()
        assert mocked_instance.status == 'failed'
        assert mocked_instance.reason == 'dummy message'
        
```

LukaszMrugala commented on scripts/tests/twister/test_handlers.py:
```diff
@@ -0,0 +1,822 @@
+
+    mock_process.communicate = mock.Mock(side_effect=raise_timeout_fn)
+
+    script = [os.path.join('test','script', 'path'), 'arg']
```
Yes, it's always the same path. Trying other `Popen` parameter versions would probably qualify as testing `subprocess` itself, which we do not want to do.

golowanow replied:
> Yes, it's always the same path. 

then we should test it just as we expect the value as a string with a path

LukaszMrugala replied:
I'm not sure I understand

golowanow commented on scripts/tests/twister/test_handlers.py:
```diff
@@ -0,0 +1,822 @@
+]
+
+@pytest.mark.parametrize("runner, product, west_flash_options, runner_options, serial_not_pty, expected", TESTDATA_8)
+def test_devicehandler_handle(mocked_instance, runner, product, west_flash_options, runner_options, serial_not_pty, expected):
```
imo, without `handle()` refactoring, this new test might become an additional hindrance and a problem to maintain it with minor changes/fixes.
m.b. we should keep this function not covered with tests now until its refactoring ?

LukaszMrugala replied:
As we do not know for know when or even if such a refactoring will come, it might still be a good idea to cover it. The DeviceHandler's `handle` is not so complex that the maintenance would be impossible, its main problem is the length and the amount of responsibilities. Because of that splitting this test Into multiple later shouldn't be that hard.

LukaszMrugala committed:
scripts: pylib: twister: twisterlib: Handlers refactor

Simple refactoring aiming to reduce the average method length.
Minor corrections of the handlers module, removing wholly unused
variables, etc.

Signed-off-by: Lukasz Mrugala <lukaszx.mrugala@intel.com>

LukaszMrugala committed:
scripts: tests: twister: Handlers test update post refactor

After the handler.py module has been split up more,
we can update the relevant tests.
100% coverage achieved.

Signed-off-by: Lukasz Mrugala <lukaszx.mrugala@intel.com>

hakehuang reviewed (commented):
can you share the test report?

LukaszMrugala commented:
> can you share the test report?

Here you can see the console outputs for:
1. Running the handlers tests by themselves
2. Running all the tests in coverage
3. Generating a textual report on coverage

```
<user>@<machine>:~/zephyr_github/os.rtos.zephyr.zephyr$ pytest scripts/tests/twister/test_handlers.py
============================== test session starts ===============================platform linux -- Python 3.10.12, pytest-7.3.1, pluggy-1.0.0
rootdir: <ZEPHYR_BASE>
collected 126 items

scripts/tests/twister/test_handlers.py ................................... [ 27%]
.......................................................................... [ 86%]
.................                                                          [100%]

============================== 126 passed in 5.53s ===============================
<user>@<machine>:~/zephyr_github/os.rtos.zephyr.zephyr$ cd scripts
<user>@<machine>:~/zephyr_github/os.rtos.zephyr.zephyr/scripts$ coverage run -m pytest $ZEPHYR_BASE/scripts/tests/twister/
============================== test session starts ===============================platform linux -- Python 3.10.12, pytest-7.3.1, pluggy-1.0.0
rootdir: <ZEPHYR_BASE>/scripts
collected 302 items

tests/twister/test_cmakecache.py ......................................    [ 12%]
tests/twister/test_environment.py ............................             [ 21%]
tests/twister/test_handlers.py ........................................... [ 36%]
.......................................................................... [ 60%]
.........                                                                  [ 63%]
tests/twister/test_harness.py ..........                                   [ 66%]
tests/twister/test_quarantine.py .............                             [ 71%]
tests/twister/test_runner.py ......                                        [ 73%]
tests/twister/test_testinstance.py .........................               [ 81%]
tests/twister/test_testplan_class.py .................................     [ 92%]
tests/twister/test_testsuite.py ...........                                [ 96%]
tests/twister/test_twister.py .......                                      [ 98%]
tests/twister/pytest_integration/test_harness_pytest.py .....              [100%]

================================ warnings summary ================================
pylib/twister/twisterlib/testplan.py:68
  <ZEPHYR_BASE>/scripts/pylib/twister/twisterlib/testplan.py:68: PytestCollectionWarning: cannot collect test class 'TestPlan' because it has a __init__ constructor (from: tests/twister/test_testplan_class.py)    class TestPlan:

tests/twister/test_testinstance.py: 84 warnings
tests/twister/test_testplan_class.py: 120 warnings
tests/twister/test_twister.py: 1 warning
  <ZEPHYR_BASE>/scripts/pylib/twister/twisterlib/config_parser.py:231: DeprecationWarning: Do not specify CONF_FILE, OVERLAY_CONFIG, or DTC_OVERLAY_FILE in extra_args. This feature is deprecated and will soon result in an error. Use extra_conf_files, extra_overlay_confs or extra_dtc_overlay_files YAML fields instead
    warnings.warn(

tests/twister/test_testinstance.py: 21 warnings
tests/twister/test_testplan_class.py: 30 warnings
tests/twister/test_twister.py: 1 warning
  <ZEPHYR_BASE>/scripts/pylib/twister/twisterlib/config_parser.py:122: DeprecationWarning: Space-separated lists are deprecated, use YAML lists instead
    warnings.warn(

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
======================= 302 passed, 258 warnings in 8.66s ========================
<user>@<machine>:~/zephyr_github/os.rtos.zephyr.zephyr/scripts$ coverage report -m
Name                                                                             Stmts   Miss  Cover   Missing
--------------------------------------------------------------------------------------------------------------
[.../tmp files...]
[.../usr/python files...]
[...dts files..]
pylib/build_helpers/domains.py                                                      78     48    38%   55-69, 78-92, 98, 101-119, 122, 125, 131-132, 136, 140, 144, 148
pylib/twister/expr_parser.py                                                       182    126    31%   17-18, 52-53, 57-58, 63-64, 92-93, 98, 111, 115, 119, 123, 134, 138, 142-143, 147, 151-152, 156, 160, 164-165, 170, 175, 178-181, 186, 189-191, 194-200, 203-287, 293-301, 311-317, 322-339
pylib/twister/parsetab.py                                                           18      0   100%
pylib/twister/scl.py                                                                34     11    68%   19-20, 42-47, 58, 62-65
pylib/twister/twisterlib/cmakecache.py                                              90      0   100%
pylib/twister/twisterlib/config_parser.py                                          116     20    83%   22, 27-30, 107, 127, 131, 135, 140, 145, 149, 154, 188, 192, 194, 196, 201, 247, 260-261
pylib/twister/twisterlib/environment.py                                            292      6    98%   28, 237-239, 850-851
pylib/twister/twisterlib/error.py                                                   11      0   100%
pylib/twister/twisterlib/handlers.py                                               662      0   100%
pylib/twister/twisterlib/harness.py                                                401    203    49%   85-108, 115-120, 128-130, 134-154, 159-165, 168-216, 234-244, 266, 272, 277-280, 284-320, 323-343, 351-365, 369-374, 381-383, 386, 400, 419, 516-558, 569-574
pylib/twister/twisterlib/jobserver.py                                              132     98    26%   23-25, 28, 31-32, 40, 45, 50, 60-65, 79-96, 103-108, 135-183, 191-202, 206-213, 217-219, 230-236
pylib/twister/twisterlib/log_helper.py                                               9      1    89%   25
pylib/twister/twisterlib/mixins.py                                                   2      0   100%
pylib/twister/twisterlib/platform.py                                                61      7    89%   75, 97-99, 104-105, 108
pylib/twister/twisterlib/quarantine.py                                              80      0   100%
pylib/twister/twisterlib/runner.py                                                 843    609    28%   35, 64-97, 100-112, 116-117, 121-122, 126-127, 131-132, 136-137, 141-142, 146-147, 151-152, 156-157, 161-162, 166-167, 171-172, 176-177, 181-182, 186-187, 191-192, 196-197, 201-202, 206-207, 211-212, 216-217, 242-243, 247-315, 319-408, 420-511, 528-542, 545-558, 562-677, 680-716, 720-752, 755-762, 803, 812, 825-826, 836, 843, 883-974, 1008-1019, 1022-1023, 1027-1049, 1052-1059, 1063-1078, 1083-1091, 1095-1165, 1173-1180, 1183, 1191-1214, 1218-1243, 1246-1266
pylib/twister/twisterlib/size_calc.py                                              208    173    17%   119-130, 133-142, 149, 156, 163-167, 174, 181, 189-191, 195-203, 209-221, 225-286, 291-293, 300-307, 315-330, 339-352, 360, 368-377, 385-397, 405-419, 426-438, 442-460
pylib/twister/twisterlib/testinstance.py                                           200     57    72%   91, 95, 100-101, 104, 107, 110-114, 131, 148, 153-185, 201, 210, 216-217, 280-281, 289, 303-305, 313-316, 319
pylib/twister/twisterlib/testplan.py                                               658    360    45%   25-26, 40, 122, 144-148, 151-163, 166-197, 201-242, 251-296, 301-302, 306-316, 319-328, 331-336, 339-366, 369-376, 380-388, 391-402, 405-406, 409, 414-415, 430-431, 434, 442-444, 463-477, 509-516, 528-529, 533-535, 539, 550-612, 646-648, 651-652, 654, 656, 664-668, 681-683, 708, 718-721, 724, 728, 731-732, 735-738, 741, 744, 747, 750, 753, 756, 759, 762, 767, 770, 773, 776, 779, 784, 787, 790, 793, 799, 802, 806, 809-811, 814, 817, 820, 830-847, 875-876, 880-881, 883-884, 887-889, 913-918, 930-931, 940-947, 958-973, 981-985
pylib/twister/twisterlib/testsuite.py                                              211     42    80%   54, 104, 122, 136-145, 220, 233-238, 269-289, 300-301, 305-309, 321, 324, 339, 342, 391, 396-397
tests/twister/conftest.py                                                           67     10    85%   22, 88-96
tests/twister/pytest_integration/test_harness_pytest.py                             83      0   100%
tests/twister/test_cmakecache.py                                                    75      0   100%
tests/twister/test_environment.py                                                  131      4    97%   146-151, 416
tests/twister/test_handlers.py                                                     713      6    99%   99, 101, 686, 873, 1694, 1918
tests/twister/test_harness.py                                                       83      0   100%
tests/twister/test_quarantine.py                                                    42      0   100%
tests/twister/test_runner.py                                                        93      0   100%
tests/twister/test_testinstance.py                                                  56      0   100%
tests/twister/test_testplan_class.py                                               165      4    98%   118, 166, 173, 207
tests/twister/test_testsuite.py                                                     24      1    96%   74
tests/twister/test_twister.py                                                       46      1    98%   50
zephyr_module.py                                                                   351    286    19%   163-170, 174-202, 206-231, 237-265, 271-283, 287-295, 299-311, 315-329, 333-353, 357-377, 382-399, 414-514, 525-527, 544-550, 553, 585-594, 598-601, 606-617, 622-625, 631-729, 733
--------------------------------------------------------------------------------------------------------------
TOTAL                                                                            15247   7689    50%
```

That should be sufficient to verify the 100% coverage on handlers. The HTML report uses this information as its base.

gchwier commented on scripts/pylib/twister/twisterlib/handlers.py:
```diff
@@ -574,6 +554,86 @@ def handle(self, harness):
+                serial_pty,
+                ser_pty_process
+            )
+        except serial.SerialException:
```
same try-except is in new-extracted method `_create_serial_connection`. Maybe better to return status in `_create_serial_connection` method and here only use `if` statement, instead of raising again same exception.
Could be also fine to extract part from the error handling section in `_create_serial_connection` to separate method and call it here (to avoid recursive exceptions)

gchwier commented on scripts/pylib/twister/twisterlib/handlers.py:
```diff
@@ -728,6 +782,51 @@ def _thread(handler, timeout, outdir, logfile, fifo_fn, pid_fn, results, harness
+            handler.instance.reason = "Unknown"
+
+    @staticmethod
+    def _thread(handler, timeout, outdir, logfile, fifo_fn, pid_fn, results,
```
Maybe as a next step could be good to extract methods related to Thread to a separate Class (instead of using `@staticmethod` and calling extracted method with `QEMUHandler._thread...`. You can check that class:
https://github.com/zephyrproject-rtos/zephyr/blob/2e333000b5c4d316224bf850449d947824a77951/scripts/pylib/pytest-twister-harness/src/twister_harness/device/fifo_handler.py

gchwier reviewed (commented):
Could be fine to add also some annotations during that refactoring... At least to `instance` definition in Handler class. It is cool, when your IDE recognizes an object, that is used everywhere.

gchwier commented on scripts/pylib/twister/twisterlib/handlers.py:
```diff
@@ -574,6 +554,86 @@ def handle(self, harness):
+            master, slave = pty.openpty()
+            try:
+                ser_pty_process = subprocess.Popen(
+                    re.split('[, ]', serial_pty),
```
ok, I see now, it is the same

gchwier reviewed (approved):
Some minor comments, but can be handled in next step of refactoring to not have to much to review

LukaszMrugala commented:
> Some minor comments, but can be handled in next step of refactoring to not have to much to review

I'd personally prefer further refactoring to occur in a separate PR, even if most of current proposals are pretty simple to implement. Currently we have here method pullouts and PEP fixes with minimal logic and structural changes and yet it's not 100% trivial to review invariability here. Better to do further changes somewhere else.
